plugins/outputs/yandex_cloud_monitoring/yandex_cloud_monitoring.go

```diff
@@ -0,0 +1,276 @@
+package yandex_cloud_monitoring
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+	"github.com/influxdata/telegraf/selfstat"
+)
+
+// YandexCloudMonitoring allows publishing of metrics to the Yandex Cloud Monitoring custom metrics
+// service
+type YandexCloudMonitoring struct {
+	Timeout            internal.Duration
+	EndpointUrl        string `toml:"endpoint_url"`
+	MetadataTokenUrl   string `toml:"metadata_token_url"`
+	MetadataFolderUrl  string `toml:"metadata_folder_url"`
+	Service            string `toml:"service"`
+	FolderID           string `toml:"folder_id"`
+	IAMTokenFromConfig string `toml:"iam_token"`
+
+	IAMToken               string
+	IamTokenExpirationTime time.Time
+
+	client *http.Client
+
+	timeFunc func() time.Time
+
+	MetricOutsideWindow selfstat.Stat
+}
+
+type yandexCloudMonitoringMessage struct {
+	TS      string                        `json:"ts,omitempty"`
+	Labels  map[string]string             `json:"labels,omitempty"`
+	Metrics []yandexCloudMonitoringMetric `json:"metrics"`
+}
+
+type yandexCloudMonitoringMetric struct {
+	Name       string            `json:"name"`
+	Labels     map[string]string `json:"labels"`
+	MetricType string            `json:"type,omitempty"` // DGAUGE|IGAUGE|COUNTER|RATE. Default: DGAUGE
+	TS         string            `json:"ts,omitempty"`
+	Value      float64           `json:"value"`
+}
+
+type MetadataIamToken struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int64  `json:"expires_in"`
+	TokenType   string `json:"token_type"`
+}
+
+const (
+	defaultRequestTimeout    = time.Second * 5
+	defaultEndpointUrl       = "https://monitoring.api.cloud.yandex.net/monitoring/v2/data/write"
+	defaultMetadataTokenUrl  = "http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token"
+	defaultMetadataFolderUrl = "http://169.254.169.254/computeMetadata/v1/instance/attributes/folder-id"
+	maxRequestBodySize       = 4000000
+)
+
+var sampleConfig = `
+  ## Timeout for HTTP writes.
+  # timeout = "20s"
+
+  ## Normally should not be changed
+  # endpoint_url = "https://monitoring.api.cloud.yandex.net/monitoring/v2/data/write"
+
+  ## Normally folder ID is taken from Compute instance metadata
+  # folder_id = "..."
+
+  ## Can be set explicitly for authentification debugging purposes 
+  # iam_token = "..."  
+`
+
+// Description provides a description of the plugin
+func (a *YandexCloudMonitoring) Description() string {
+	return "Send aggregated metrics to Yandex.Cloud Monitoring"
+}
+
+// SampleConfig provides a sample configuration for the plugin
+func (a *YandexCloudMonitoring) SampleConfig() string {
+	return sampleConfig
+}
+
+// Connect initializes the plugin and validates connectivity
+func (a *YandexCloudMonitoring) Connect() error {
+	if a.Timeout.Duration == 0 {
```

Comment:
`if a.Timeout.Duration <= 0 {` would be better.

Reply:
Sure